Fix return type of git_blob_is_binary()

nalimilan opened the pull request:
This triggered a failure in the libgit2 blobs test on i686 when building
RPM nightlies. I don't understand why this error does not happen on CI.

This code was introduced in 78133def0e493ae3c40c38d53a9b7634254d78fd.

Cc: @kshyatt

nalimilan committed:
Fix return type of git_blob_is_binary()

This triggered a failure in the libgit2 blobs test on i686 when building
RPM nightlies. I don't understand why this error does not happen on CI.

tkelman commented:
What version of libgit2 was that building against? I don't know whether this API has changed at all, but what was the failure you were seeing?

nalimilan commented:
I've checked the libgit2 code, and that function hasn't been touched for three years. But indeed I was using 0.24.6, so that might have revealed the bug for some obscure reason while it was hidden with 0.25.

nalimilan commented:
Error was:
```
Error in testset libgit2:
Test Failed
  Expression: LibGit2.isbinary(blob)
Error in testset libgit2:
Error During Test
  Got an exception of type MethodError outside of a @test
  MethodError: no method matching split(::Ptr{Void}, ::String)
  Closest candidates are:
    split{T<:SubString}(!Matched::T<:SubString, ::Any; limit, keep) at strings/util.jl:231
    split{T<:AbstractString}(!Matched::T<:AbstractString, ::Any; limit, keep) at strings/util.jl:254
  Stacktrace:
   [1] #wrapped_lines#175(::Int32, ::Int32, ::Function, ::String) at ./markdown/render/terminal/formatting.jl:68
   [2] #wrapped_lines#175(::Int32, ::Int32, ::Function, ::String) at ./markdown/render/terminal/formatting.jl:67
   [3] VersionNumber(::Int32, ::Int32, ::Int32, ::Tuple{}, ::Tuple{}) at ./version.jl:38 (repeats 2 times)
ERROR: LoadError: 
```
https://copr-be.cloud.fedoraproject.org/results/nalimilan/julia-nightlies/fedora-26-i386/00504417-julia/build.log.gz

tkelman commented:
Looks like you found 2 bugs then, and you're only fixing the easy one! `split(content(blob), "\n")` isn't going to work if `content(blob)` returns a `Ptr{Void}`. The `show(::IO, ::GitBlob)` was only being tested for binary `GitBlob`s (aside: funny naming on libgit2's part, I was under the impression that the first b in blob meant binary), you just happened to hit that code path in your tests by accident before the CI here happened to do so. I do wonder what's different about your environment than the CI here that would cause an incorrect ccall return type to behave differently, or maybe it's intermittent.

nalimilan commented:
Indeed. I'd appreciate if somebody familiar with libgit2 could fix the second bug, as I don't really now what are the different types of `GitBlob`s.

FWIW, the failure was pretty systematic on the Copr.

simonbyrne commented:
The `content` issue is fixed by #20104 (see https://github.com/JuliaLang/julia/pull/20104/files#diff-3384e375f713976e7e985a7dfe83a626R3).

kshyatt commented:
So I don't have to do anything? Thanks for finding and fixing this!

tkelman commented:
Would be good to dig into why only you were hitting this.

nalimilan commented:
Hard to tell, but I'm the only one to use 0.24 and it doesn't sound impossible that the 32 additional bytes happened to be all zeros on 0.25, which most people use. Bugs involving bools are less likely to be noticed than those involving types in which all 64-bits are actually significant.
